mpl2: IOs abstraction adaptation to ORFS flow - first version (simplified)

AcKoucher opened the pull request:
Resolve the first top level task in #5669.
Resolve #6050.
Resolve #4867.

### Introductory Considerations
- This is the first step of the enhancements to adapt the IOs abstractions inside mpl2 to ORFS flow.
- This first version will not consider the dimensions of the constraints whether they are -exclude constraints or pins' constraint regions.

### Goals
1. Improve QoR in terms of WL.
2. Ease the life of the annealer for some cases (e.g., sky130hd/uW) that require extreme virtual std cell utilization configurations to converge.
3. Further steps into switching from manual mpl to mpl2 for cases such as sky130hd/chameleon (This will still require clustering work)

### Main Changes

#### Modify IO Clusters
IO clusters that represent bundledIOs created from the pins placed by random PPL are now groups of unplaced IOs with the same constraints based on the following:
- Individual Pin Constraints:
  - If an IO has a constraint region in a certain boundary, it is constrained to that entire boundary.
  - If an IO has no constraint region, it is constrained to all boundaries.
 
- Global Pins' Constraints (-exclude)
  - If a boundary has more than 70% of its extension blocked, it is considered a blocked boundary for IOs.

#### What Annealing Sees
- WL between a cluster/macro (soft/hard) M and an IO cluster IO is represented as follow:
  - If IO represents pins constrained to a certain boundary A, WL is the minimum distance between M's bundled center pin and A.
  - If IO represents unconstrained pins, WL is the minimum distance between M's bundled center pin and the closest non-blocked boundary.

**Obs: This distance only reduces cost when M is inside the outline.**

### Collateral Changes

#### Needed for the Approach to Work
1. Add Cluster support to HardMacro so that SA can check if they are IO clusters' fixed terminals.
2. Make both Soft/HardMacroSACores have access to the PhisycalHierarchical data where the global constraints are now stored.
3. Make fixed terminals that represent IO clusters have the cluster data inside them so that this data is accessible inside SA for both Soft and Hard. I.e., the cluster of a fixed terminal that represents an IO cluster will be != nullptr

#### Blockages for Pin Access
Generated based on a combination of both global and individual pins' constraints:
1. If there are pins constrained to a certain boundary, we'll generate a blockage for that boundary.
2. If there are no pins constrained a boundary A but there are pins constrained to all boundaries, A won't have a pin access blockage. (Specially important so that SA has more freedom).
3. The depth of the blockage is a combination of the extension of IO clusters and std cell area (_this can definitely be further improved to consider the IOs density_)

#### Orientation Improvement
When iterating a net, compute HPWL based on:
1. If a pin has a constraint region, use its center.
2. If a pin doesn't have a constraint region, use the closest point of the closest boundary to that macro pin.

#### Debug Mode
1. Blocked boundaries are marked with a red X in the center of their extension
2. Draw distance to constraint || closest boundary only when the cluster/macro is inside the outline

eder-matheus commented on src/mpl2/src/SimulatedAnnealingCore.cpp:
```diff
@@ -272,10 +295,18 @@ void SimulatedAnnealingCore<T>::calWirelength()
+    T& source = macros_[net.terminals.first];
+    T& target = macros_[net.terminals.second];
+
+    if (target.isIOCluster()) {
```
Will `source` never be a macro?

AcKoucher replied:
Afaik, no. Perhaps add a critical for that case?

eder-matheus replied:
I don\t think it is necessary. I was just curious why you're only checking if the `target` is an IO cluster, but if source is never an IO cluster, that's fine.
(My original comment was wrong, btw)

eder-matheus commented on src/mpl2/src/hier_rtlmp.cpp:
```diff
@@ -914,142 +912,119 @@ void HierRTLMP::setTightPackingTilings(Cluster* macro_array)
+                                        const Rect& die)
+{
+  if (constraint_boundary == L) {
+    const Rect left_io_blockage(
```
Similar to another comment I made, in the future we should work on enhancing this process to consider the actual constraint region. Blocking the whole edge might be to extreme for cases where the constraint region is small.

AcKoucher replied:
I''ll open an issue for that.

AcKoucher replied:
Opened #5816

AcKoucher commented on src/mpl2/src/clusterEngine.cpp:
```diff
@@ -327,165 +329,155 @@ void ClusteringEngine::setBaseThresholds()
+  return blockage_extension_map;
+}
+
+void ClusteringEngine::setIOClusterDimensions(const odb::Rect& die,
```
wouldn't "find" fit better for a getter? It looks to me that we're literally setting the dimension inside this method.

eder-matheus replied:
I was thinking on only using `set` for class member setters, but it seems a bit restrictive. I'm fine with either set or find.

AcKoucher commented on src/mpl2/src/SimulatedAnnealingCore.cpp:
```diff
@@ -288,6 +319,70 @@ void SimulatedAnnealingCore<T>::calWirelength()
+        += net_weight
+           * std::min(
+               {dist_to_left, dist_to_right, dist_to_bottom, dist_to_top});
+  } else if (constraint_boundary == Boundary::L
```
It's the first option and this is wrong and should be fixed.

maliberty commented:
"Lint Tcl code" needs fixing.

If a pin has been manually placed with place_pin (status=FIRM), will this code consider its location rather than its constraint?

AcKoucher commented:
@maliberty I edited the description of the PR/Issue to make it clear that this is a simplified version of these enhancements.
I also pushed the change regarding the blocked boundaries as discussed.

When addressing the review I'll fix the Lint Tcl.

AcKoucher commented:
Resolved.

maliberty commented:
Some unit test need fixing:
```
[2024-11-14T19:15:30.551Z] mpl2
[2024-11-14T19:15:37.996Z] io_constraints1 (tcl) *FAIL*
[2024-11-14T19:15:46.470Z] io_constraints2 (tcl) *FAIL*
```

maliberty commented:
It would be good to merge master and do a fresh secure CI as this PR is a bit old

maliberty commented on src/mpl2/src/hier_rtlmp.cpp:
```diff
@@ -923,142 +921,119 @@ void HierRTLMP::setTightPackingTilings(Cluster* macro_array)
+                                        const float depth,
+                                        const Rect& die)
+{
+  if (constraint_boundary == L) {
+    const Rect left_io_blockage(
+        die.xMin(), die.yMin(), die.xMin() + depth, die.yMax());
     boundary_to_io_blockage_[L] = left_io_blockage;
     macro_blockages_.push_back(left_io_blockage);
   }
```
This would be more succinct as
```
Rect blockage = die;
if (constraint_boundary == L) {
  blockage.set_xhi(die.xMin() + depth);
} else if (..) {...}...

 boundary_to_io_blockage_[constraint_boundary] = blockage;
 macro_blockages_.push_back(blockage);
```

AcKoucher replied:
This is mpl2::Rect so it doesn't have set_xhi. I'll add it.

maliberty reviewed (commented):
Pending updated secure CI.

AcKoucher commented:
I ran an updated Secure-CI in the weekend and the only concerning thing is a WL degradation in gf180/uartblocks:

gf180 uart-blocks
      [ERROR] globalroute__antenna_diodes_count fail test: 11.0 <= 10.0
      [ERROR] detailedroute__route__wirelength fail test: 23164.0 <= 19722.0
      
I'll investigate it.

AcKoucher commented:
The problem with uart-blocks was due to the fact that I wasn't considering the case in which a pin was not constrained to any region in the orientation improvement step. Now we compute WL with a similar technique used in SA: we get the closest point in the closest boundary and merge it's rect in the net_box.

Before:
<img src="https://github.com/user-attachments/assets/a7927873-ad43-4030-83cc-b1fe80114689" width="300">

After:
<img src="https://github.com/user-attachments/assets/70285583-b7ff-460e-ab1b-dc67d39e4171" width="300">

Also:
While updating the regression tests I found a bug in the pin access blockage creation which was fixed.
The same for a bug in graphics.

AcKoucher commented:
I'm running a new secure-CI with my latest changes and master.

eder-matheus reviewed (approved):
I don't have further comments, but I'll wait for @maliberty's final review before merging.

maliberty commented:
> There a couple designs that need metrics update. 
Is there a PR for this?

AcKoucher commented:
> Is there a PR for this?

I opened [#2582](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/2582). However, a metrics conflict emerged from another PR. I'm re-running Secure-CI to update the metrics correctly.

maliberty commented:
@AcKoucher what remains to merge this?

AcKoucher commented:
@maliberty This PR is blocked by a gpl divergence in uW (See #6235).

AcKoucher commented:
I resolved the conflicts here.

With the updates from master, GPL no longer diverges in uW. However, now GRT is throwing an error due to high congestion in post-drt. I'm investigating.

AcKoucher commented:
Updating..

I ran a new Secure-CI with a custom GRT routing layer adjustment and it solved uW's congestion problem.
However, now g12/ca53 is failing in DPL.

AcKoucher commented:
Settings changes and metrics update in [#2706](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/2706)
